### PR TITLE
cups: enable compiling with musl

### PR DESCRIPTION
**Pull request by cdepillabout**
This PR enables `cups` to be able to be compiled with `pkgsMusl`.

`cups` can't depend on `systemd` when being compiled with `pkgsMusl`, because it is not possible to build `systemd` with `pkgsMusl`.

###### Motivation for this change

This is for https://github.com/nh2/static-haskell-nix/issues/50.  I'd like to get GTK compiling with `pkgsMusl`, and GTK has a dependency on `cups`.

This can be tested like the following:

```console
$ nix-build -A pkgsMusl.cups
...
/nix/store/mw3v16g8qbyr10dwz18krpssilpap67c-cups-2.2.12
```

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @nh2 @domenkozar @matthewbauer

### CODE REVIEW

**Comment by cdepillabout** [comment]
Note that currently there are 3 commits in this PR.  However, two of them are from other PRs, and shouldn't be included when this is merged in.

This PR should only be merged in after the following two have been merged in:

- #71560 (4845af0) enable building dbus with musl.  dbus is a dependency of cups.
- #71563 (b1b809f) enable building libusb1 with musl.  libusb1 is a dependency of cups.

The only commit that is relevant to this PR (well, `cups`), is 317cd4d.

**Comment by cdepillabout** [comment]
@GrahamcOfBorg build cups

@GrahamcOfBorg build pkgsMusl.cups

**Comment by cdepillabout** [comment]
This is related to the following three PRs:

- dbus: enable building with musl #71560 
- libusb1: enable compiling with musl #71563 
- vte: enable compiling with musl #71577

**Comment by matthewbauer** [comment]
Systemd /should/ build on Musl right? Perhaps some patch exists for it?

**Comment by cdepillabout** [comment]
@matthewbauer As far as I am aware, it is difficult to get systemd to build with musl:

https://github.com/NixOS/rfcs/blob/master/rfcs/0023-musl-libc.md#big-incompatibilities

I think @nh2 has done a lot of work getting stuff building in `pkgsMusl`.  Maybe he can say something about systemd?

**Comment by flokli** [comment]
Feedback to things like https://lists.freedesktop.org/archives/systemd-devel/2014-September/023177.html was mostly constructive IMHO.

I think in general systemd upstream seems to be fine with accepting patches enhancing musl compat, it might be just more complicated if it's about things not provided by the libc, if there's different opinions on whether it should be in there or not.

Another thread from `systemd/casync`: https://github.com/systemd/casync/issues/129#issuecomment-362842473

**Comment by flokli** [comment]
That being said, removing systemd in case of building with musl might be a good workaround for now, but having a musl-based systemd might become interesting for NixOS-on-whatever-device setups :-)

**Comment by cdepillabout** [comment]
This has the ` 10.rebuild-linux: 2501-5000 ` tag.  Would it be better to go into staging?

File: pkgs/misc/cups/default.nix
Comment (nh2) [inline pkgs/misc/cups/default.nix]: Same comment as https://github.com/NixOS/nixpkgs/pull/71563#discussion_r337311093

**Review by nh2 — CHANGES_REQUESTED** [review]

**Comment by nh2** [comment]
On the point of systemd + musl compatibility:

Upstream support for making systemd support building on musl is mixed.

systemd maintainers are mostly of the opinion that they should use glibc functions they consider useful, as opposed to re-implementing them (code duplication) to support other libcs.

I think one could say they consider glibc a more of a "normal library dependency" that provides the functions they want, and not a "minimal portable standard".

On the functions that systemd uses that glibc has and musl does not, the opinions diverge on whether they are are useful or good to have in a libc.

I personally forecast that systemd and musl slowly get together over time, but don't expect that this is very soon, so I think the approach in this PR to disable it for now is exactly the right choice.

**Comment by nh2** [comment]
@cdepillabout Another request across your 4 PRs (not a blocker but a nice-to-have):

Could you make `enableSystemd ? (stdenv.isLinux && !stdenv.hostPlatform.isMusl)` a _parameter_ (top-level argument) of each package?

That would allow us to easily turn off systemd deps even when using glibc, which would be convenient to quickly check whether things are fine with glibc in situations where the musl builds are broken. It would also be in line with how other packages are written.

File: pkgs/misc/cups/default.nix
Comment (cdepillabout) [reply pkgs/misc/cups/default.nix]: This has been done in 3a4584c.

**Comment by cdepillabout** [comment]
> Could you make enableSystemd ? (stdenv.isLinux && !stdenv.hostPlatform.isMusl) a parameter (top-level argument) of each package?

This has been done in 3a4584c.  Thanks for the suggestion!

**Review by nh2 — APPROVED** [review]

**Comment by cdepillabout** [comment]
@GrahamcOfBorg build pkgsMusl.cups

**Comment by nh2** [comment]
> This has the `10.rebuild-linux: 2501-5000` tag. Would it be better to go into staging?

I've force-pushed a slightly reworked version of the patch that keeps the original order of `buildInputs`, so that on non-musl, this should not create any changes.

**Comment by cdepillabout** [comment]
@nh2 Didn't realize this was the thing forcing the mass rebuild!  Thanks for catching and fixing this!

**Comment by nh2** [comment]
Rebased on master

**Comment by nh2** [comment]
This has worked fine for me in static-haskell-nix so far, and no changes for non-musl; merging.

**Comment by cdepillabout** [comment]
@nh2 Thanks for fixing up these four PRs and merging them in :-)

**Pull request merged**
